Libraries/Cordio/wsf/sources/targets/freertos/wsf_nvm.c

```diff
@@ -373,3 +423,65 @@ void WsfNvmEraseDataAll(WsfNvmCompEvent_t compCback)
     compCback(TRUE);
   }
 }
+
+/*************************************************************************************************/
+/*!
+ *  \brief  Defragment NVM
+ *  \param  copyBuf       Scratch buffer to temporarily copy NVM stored data.
+ *  \param  size          Size of copyBuf in bytes.
+ *  \return  TRUE if defragment successful. FALSE otherwise
+ * 
+ *  \note   Defragment should only be called when the storage is full and a record has been invalidated.
+ *          Furthermore, copyBuf must be at least the size of WSF NVM allocated flash.
+ */
+/*************************************************************************************************/
+bool_t WsfNvmDefragment(uint8_t *copyBuf, uint32_t size)
+{
+#if WSF_ASSERT_ENABLED == 1
+    WSF_ASSERT(copyBuf && size >= wsfNvmCb.totalSize);
+#else
+    if (!copyBuf || size < wsfNvmCb.totalSize) {
+        WSF_TRACE_INFO0("Not enough memory given to defragment NVM");
+        return FALSE;
+    }
+#endif
+
+    WsfCsEnter();
+
+    bool_t defragPossible = FALSE;
+    uint32_t currentOffset = WSF_NVM_START_ADDR;
+    uint32_t copyOffset = 0;
+    WsfNvmHeader_t header = { 0 };
+
+    /* Get the first header*/
+    PalFlashRead(&header, WSF_NVM_HEADER_SIZE, currentOffset);
+    while (currentOffset < wsfNvmCb.availAddr && header.id != WSF_NVM_UNUSED_FILECODE) {
+        const uint32_t fileSize = WSF_NVM_FILE_SIZE(header.len);
+
+        /*If the entry is valid copy it into the defragmentation buffer, if its reserved we can defrag it*/
+        if (header.id != WSF_NVM_RESERVED_FILECODE) {
+            PalFlashRead(&copyBuf[copyOffset], fileSize, currentOffset);
```

Comment:
I am getting a HardFault on the `memcpy` call inside this call.

I have defined `PAL_NVM_SIZE` to be 0x2000. I also fixed the bug in [this comment](https://github.com/analogdevicesinc/msdk/issues/798#issuecomment-2038424644).

I filled up the DB, and now when this code executes, `wsfNvmCb.availAddr`'s value is 0x2010, which is past the bounds of the DB. So the `header` actually has invalid data. `header.len` is 0x2000242c, which ultimately causes the HardFault in `memcpy`.

Reply:
Can you see where the address gets set to the out of bounds state? Once again the ```wsfHaveEnoughSpace``` function is intended to block against this condition.  If it isn't, I need to fix that. A workaround that I don't like is just getting the min between the available address and the total size

Reply:
It's being set in `WsfNvmInit` on a subsequent boot, after the failed write due to insufficient space.